### PR TITLE
fixbug:permission unlocking issue resolved

### PR DESCRIPTION
**Pull request by ShenDing1125**
- permission unlocking issue resolved
- fixing multiple request loading issue

### CODE REVIEW

**Comment by ShenDing1125** [comment]
[(2)【課程】同一單元會一下跳購買方案解鎖內容，一下又跳回課程影片](https://trello.com/c/E0RTsjIy/5143-2%E3%80%90%E8%AA%B2%E7%A8%8B%E3%80%91%E5%90%8C%E4%B8%80%E5%96%AE%E5%85%83%E6%9C%83%E4%B8%80%E4%B8%8B%E8%B7%B3%E8%B3%BC%E8%B2%B7%E6%96%B9%E6%A1%88%E8%A7%A3%E9%8E%96%E5%85%A7%E5%AE%B9%EF%BC%8C%E4%B8%80%E4%B8%8B%E5%8F%88%E8%B7%B3%E5%9B%9E%E8%AA%B2%E7%A8%8B%E5%BD%B1%E7%89%87)

File: src/pages/ProgramContentPage/ProgramContentBlock.tsx [hunk 1]
```diff
@@ -372,32 +391,51 @@ const useHasProgramContentPermission: (
   programContentId: string,
 ) => {
   hasProgramContentPermission: boolean
+  isLoading: boolean
 } = (programId, programContentId) => {
   const { currentMemberId, authToken } = useAuth()
   const [data, setData] = useState<{ programContentId: string } | {}>({})
```
Comment (Ponchimeow) [inline src/pages/ProgramContentPage/ProgramContentBlock.tsx:397]: this state is not used

File: src/pages/ProgramContentPage/ProgramContentBlock.tsx [hunk 2]
```diff
@@ -372,32 +391,51 @@ const useHasProgramContentPermission: (
...
+            params: { memberId: currentMemberId },
+            headers: { authorization: `Bearer ${authToken}` },
+            cancelToken: source.token,
+          })
 
-        setData(data)
-      } catch (err) {
-        console.log(err)
+          if (!isUnmounted.current) {
+            setData(data)
```
Comment (Ponchimeow) [inline src/pages/ProgramContentPage/ProgramContentBlock.tsx:414]: `setData` could be remove

Comment (ShenDing1125) [reply src/pages/ProgramContentPage/ProgramContentBlock.tsx]: Thank you for your assistance

File: src/pages/ProgramContentPage/ProgramContentBlock.tsx [hunk 3]
```diff
@@ -209,6 +222,12 @@ const ProgramContentBlock: React.VFC<{
               <LockIcon className="mr-2 mb-1" />
               {formatMessage(pageMessages.ProgramContentBlock.loginTrial)}
             </Button>
+          ) : isLoading ? (
+            !showSkeleton && (
+              <div style={{ width: '80%', margin: '0 auto 1.75rem' }}>
+                <SkeletonText mt="4" noOfLines={4} spacing="4" skeletonHeight="2" />
+              </div>
+            )
```
Comment (Ponchimeow) [inline src/pages/ProgramContentPage/ProgramContentBlock.tsx:230]: loading status should be earlier than the data layer.

File: src/pages/ProgramContentPage/ProgramContentBlock.tsx [hunk 4]
```diff
@@ -372,32 +391,49 @@ const useHasProgramContentPermission: (
...
+          }
+        } catch (err) {
+          !axios.isCancel(err) && console.log(err)
+        }
       }
-    }
-  }, [currentMemberId, programContentId, programId])
+    },
+    [currentMemberId, programContentId, programId],
+  )
```
Comment (Ponchimeow) [inline src/pages/ProgramContentPage/ProgramContentBlock.tsx:422]: Is the 'isUnmounted' parameter necessary?
```
          await axios.get(`${process.env.REACT_APP_LODESTAR_SERVER_ENDPOINT}${route}`, {
            params: { memberId: currentMemberId },
            headers: { authorization: `Bearer ${authToken}` },
            cancelToken: source.token,
          }).then(({ data }) => {
            setHasProgramContentPermission(Object.keys(data).length > 0)
            setIsLoading(false)
          })
```

File: src/pages/ProgramContentPage/ProgramContentBlock.tsx [hunk 5]
```diff
@@ -104,15 +104,28 @@ const ProgramContentBlock: React.VFC<{
...
+    )
+  }, [hasProgramContentPermission, currentMemberId, isAuthenticated, programContent?.displayMode])
+
+  useEffect(() => {
+    const timer = setTimeout(() => {
+      setShowSkeleton(false)
+    }, 750)
+
+    return () => clearTimeout(timer)
+  }, [])
```
Comment (Ponchimeow) [inline src/pages/ProgramContentPage/ProgramContentBlock.tsx:128]: 覺得 statsu 與 effect 應該是不需要
isLoading 與 hasProgramContentPermission 應該就能決定要不要顯示讀取狀態或是鎖頭?

File: src/pages/ProgramContentPage/ProgramContentBlock.tsx (hunk 3 above)
Comment (ShenDing1125) [reply src/pages/ProgramContentPage/ProgramContentBlock.tsx:230]: The loading screen here replaces the original padlock, meaning that the loading screen will appear only when the user needs to wait for verification.

File: src/pages/ProgramContentPage/ProgramContentBlock.tsx (hunk 4 above)
Comment (ShenDing1125) [reply src/pages/ProgramContentPage/ProgramContentBlock.tsx:422]: Although there is a mechanism for canceling requests, this judgment is for even more secure prevention, ensuring that other actions are not executed after the component has already been destroyed.

File: src/pages/ProgramContentPage/ProgramContentBlock.tsx (hunk 5 above)
Comment (ShenDing1125) [reply src/pages/ProgramContentPage/ProgramContentBlock.tsx]: 好的，已改成經由loadingPermission來判斷是否已加載完成。

**Review by Ponchimeow — APPROVED** [review]

**Pull request merged**